src/value/owned.rs

```diff
@@ -187,52 +197,113 @@ impl PartialEq for Sequence {
 
 impl Eq for Sequence {}
 
+// This collection is broken out into its own type to allow instances of it to be shared with Rc.
+#[derive(Debug)]
+struct Fields {
+    // Key/value pairs in the order they were inserted
+    by_index: Vec<(Symbol, Element)>,
+    // Maps symbols to a list of indexes where values may be found in `by_index` above
+    by_name: HashMap<Symbol, IndexVec>,
+}
+
+impl Fields {
+    /// Gets all of the indexes that contain a value associated with the given field name.
+    fn get_indexes<A: AsSymbolRef>(&self, field_name: A) -> Option<&IndexVec> {
+        match field_name.as_symbol_ref().text() {
+            // If the provided field name symbol has undefined text...
+            None => {
+                // ...then build a cheap, stack-allocated `Symbol` that represents unknown text
+                let symbol = Symbol::unknown_text();
+                // ...and use the unknown text symbol to look up matching field values
+                self.by_name.get(&symbol)
+            }
+            Some(text) => {
+                // Otherwise, look it up by text
+                self.by_name.get(text)
+            }
+        }
+    }
+
+    /// Iterates over the values found at the specified indexes.
+    fn get_values_at_indexes<'a>(&'a self, indexes: &'a IndexVec) -> FieldValuesIterator<'a> {
+        FieldValuesIterator {
+            current: 0,
+            indexes: Some(indexes),
+            by_index: &self.by_index,
+        }
+    }
+
+    /// Gets the last value in the Struct that is associated with the specified field name.
+    ///
+    /// Note that the Ion data model views a struct as a bag of (name, value) pairs and does not
+    /// have a notion of field ordering. In most use cases, field names are distinct and the last
+    /// appearance of a field in the struct's serialized form will have been the _only_ appearance.
+    /// If a field name appears more than once, this method makes the arbitrary decision to return
+    /// the value associated with the last appearance. If your application uses structs that repeat
+    /// field names, you are encouraged to use [get_all] instead.
+    fn get_last<A: AsSymbolRef>(&self, field_name: A) -> Option<&Element> {
```

Comment:
Curious why this is `get_last` instead of `get_first` (just `get_any`). My only concern is that `get_last` forces the implementation to read and map the entire struct. That may be unproblematic now, but I don't see a good reason to choose last over first or any.

Reply:
I guess my assumption is that we would want to build the map (and the indexing to it) lazily. That doesn't seem to be happening now (and I don't think it needs to now). I confess that reviewing rust code in github is still relatively new to me, so please let me know what I'm missing.

Reply:
> My only concern is that get_last forces the implementation to read and map the entire struct. ... I guess my assumption is that we would want to build the map (and the indexing to it) lazily.

This is a good instinct, and a `LazyElement` is something we're planning to implement in the future. As you mention, this implementation (`Element`/`Struct`) is fully materialized in memory, so the choice of first or last is truly arbitrary. 

This particular method (`get_last`) lives on a private type (`Fields`) that isn't exposed to the end user. It's invoked by a public method called `get` that's defined on the `IonStruct` trait. If we were going to change the contract, that's the method we'd want to change.

At present, [`get` promises to return the last field](https://github.com/amzn/ion-rust/blob/4c7fc62e452e285862c181494267d86209a947a0/src/value/mod.rs#L483-L484) it found in the serialized struct. From an Ion data model perspective, this isn't really correct; struct fields don't have an order. The most correct thing to do would be to only offer a `get_all()` method that returned an iterator whose iteration order was not guaranteed. However, nearly all `struct` use cases don't use repeated fields and users expect to be able to access the one-and-only value for a given field trivially, so `get` is available as a convenience. If it turns out someone wants the _first_ appearance of a given field, they can call `self.get_all(name).first()` and (now that GATs are enabled) that will be basically the same cost as `get`.